src/nominal-lcf/elaborator.fun

```diff
@@ -151,50 +164,60 @@ struct
+         | _ => raise Fail ("Expected tactic, got: " ^ DebugShowAbt.toString t)
     end
 
-  and elaborateM sign rho T (us, mt) =
+  and elaborateMultitactic sign rho m : Refiner.nmultitactic =
     let
-      val mt' = evalOpen sign mt handle _ => mt
+      val (m', tau) = infer (evalOpen sign m handle _ => m)
+      val _ = case tau of MTAC => () | _ => raise Fail "elaborateMTac called on wrong sort"
     in
-      case out mt' of
-           LCF ALL $ [_ \ t'] =>
-             (fn alpha => fn jdg =>
-               let
-                 val (alpha', modulus) = probe alpha
-                 val st = T alpha' jdg
-                 val beta = prepend us (bite (!modulus) alpha)
-               in
-                 MT.ALL (elaborate sign rho t' beta) st
-               end)
-         | LCF EACH $ [_ \ v] =>
+      case m' of
+           LCF ALL $ [_ \ t] =>
+             extendTactic (elaborateStmt sign rho t)
+         | LCF EACH $ [_ \ vec] =>
              let
-               val Ts = List.map (elaborate sign rho) (elaborateVec (evalOpen sign v))
+               val Ts = List.map (elaborateStmt sign rho) (elaborateVec (evalOpen sign vec))
              in
-               fn alpha => fn jdg =>
-                 let
-                   val (alpha', modulus) = probe alpha
-                   val st = T alpha' jdg
-                   val beta = prepend us (bite (!modulus) alpha)
-                 in
-                   MT.EACH' (List.map (fn T => T beta) Ts) st
-                 end
+               fn alpha =>
+                 MT.EACH' (List.map (fn T => T alpha) Ts)
              end
-         | LCF (FOCUS i) $ [_ \ t'] =>
-             (fn alpha => fn jdg =>
-               let
-                 val (alpha', modulus) = probe alpha
-                 val st = T alpha' jdg
-                 val beta = prepend us (bite (!modulus) alpha)
-               in
-                 MT.FOCUS i (elaborate sign rho t' beta) st
-               end)
-         | _ => raise Fail ("Expecting multitac but got " ^ DebugShowAbt.toString mt')
+         | LCF (FOCUS i) $ [_ \ t] =>
+             MT.FOCUS i o elaborateStmt sign rho t
+         | _ => raise Fail "Unrecognized multitactic"
     end
 
+  and elaborateStmt sign rho : abt -> Refiner.ntactic =
+    let
+      fun go t =
+        let
+          val (t', tau) = infer (expandHypVars (evalOpen sign t handle _ => t))
+          val _ = case tau of TAC => () | _ => raise Fail "elaborateStmt called on wrong sort"
+        in
+          case t' of
+               LCF (SEQ taus) $ [([],[]) \ m, (us, _) \ t] => (us, elaborateMultitactic sign rho m) :: go t
+             | _ => [([], extendTactic (elaborateTactic sign rho t))]
+        end
+    in
+      contractTactic
+        o List.foldr
+            (fn ((us, front : Refiner.nmultitactic), rest : Refiner.nmultitactic) => fn alpha => fn st =>
+              let
+                val beta = prepend us alpha
+                val (beta', modulus) = probe (prepend us beta)
```

Comment:
nice thing about this change is that the core conceit (the probing for the modulus of continuity) is now consigned to a *single point* in the code, right here.